### PR TITLE
Add Search Impressions to EC

### PR DESCRIPTION
**Pull request by dlafreniere**
still need to add the integration test, but would like some reviews

### CODE REVIEW

File: src/plugins/ec.ts [hunk 1]
```diff
@@ -32,10 +32,29 @@ export interface ProductProperties {
...
+    category?: string;
+    variant?: string;
+    position?: number;
+    price?: number;
+}
+
+export type Impression = RequireAtLeastOne<ImpressionProperties, 'id' | 'name'>;
+export type BaseImpression = Omit<Impression, 'list'>;
+export interface ImpressionList {
+    listName?: string;
```
Comment (francoislg) [inline src/plugins/ec.ts]: 🤔 c'est tu trop confusing juste `name` ici ?

File: src/plugins/ec.ts [hunk 2]
```diff
@@ -111,9 +135,29 @@ export class EC {
...
+            const {list: listName, ...baseImpression} = impression;
+            const index = lists.findIndex(list => list.listName === listName);
+            if (index === -1) {
+                lists.push({listName: listName, impressions: [baseImpression]});
+            } else {
+                lists[index].impressions.push(baseImpression);
+            }
+            return lists;
+
+        }, new Array<ImpressionList>());
```
[the remark this answers is not in the document]
Comment (dlafreniere) [reply src/plugins/ec.ts]: ah it was the `as Impression[]` that I was missing!

File: src/plugins/ec.ts [hunk 3]
```diff
@@ -111,9 +135,29 @@ export class EC {
             };
         }, {});
 
+        const impressionsByList = this.impressions.reduce((lists, impression) => {
+            const {list: listName, ...baseImpression} = impression;
+            const index = lists.findIndex(list => list.listName === listName);
```
[the remark this answers is not in the document]
Comment (dlafreniere) [reply src/plugins/ec.ts]: at first I thought I would need the index, but that is handled later with the reduce

File: src/plugins/ec.ts (hunk 1 above)
Comment (dlafreniere) [reply src/plugins/ec.ts]: je pensais que oui parce que Impression a la base a un `name`, mais je ne suis pas contre

File: src/plugins/ec.spec.ts [hunk 4]
```diff
@@ -36,78 +36,179 @@ describe('EC plugin', () => {
...
-            ...defaultResult,
-            'pr1nm' : '🍟',
-            'pr1pr': 1.99,
-            'pr2nm' : '🍿',
-            'pr2pr': 3,
-            'pr3nm' : '🥤' ,
-            'pr3pr': 2,
+            const result = executeRegisteredHook(ECPluginEventTypes.event, {});
+
+            expect(result).not.toEqual({});
```
Comment (toussaints) [inline src/plugins/ec.spec.ts]: can't we expect what it'll be instead of what it shouldn't be?

File: src/plugins/ec.ts [hunk 5]
```diff
@@ -32,10 +32,29 @@ export interface ProductProperties {
 
 export type Product = RequireAtLeastOne<ProductProperties, 'id' | 'name'>;
 
+export interface ImpressionProperties {
+    id?: string;
```
Comment (toussaints) [inline src/plugins/ec.ts]: noob question mais pourquoi tout est optional sur ça?

File: src/plugins/ec.spec.ts (hunk 4 above)
Comment (dlafreniere) [reply src/plugins/ec.spec.ts]: I just indented this, but there is already a test that is asserting the format is okay. The goal of this test is the 2nd assertion

File: functional/ec-events.spec.ts [hunk 6]
```diff
@@ -203,6 +203,63 @@ describe('ec events', () => {
...
+            'position': 2
+        };
+
+        await coveoua('ec:addImpression', productImpression1);
+        await coveoua('ec:addImpression', productImpression2);
+        await coveoua('send', 'pageview');
+
+        const [event] = getParsedBody();
+
+        expect(event).toEqual({
```
Comment (dlafreniere) [inline functional/ec-events.spec.ts]: Validated against ca.js, the only difference was a non-standard field `il1pi1listindex: 0,` which doesn't add anything (its baked into the props)

File: src/plugins/ec.ts (hunk 5 above)
Comment (dlafreniere) [reply src/plugins/ec.ts]: id et nom sont là en option (pour que tu puisses spécifier les deux) mais, le `RequireAtLeastOne` va forcer que t'ailles au moins un des deux (tel que le spec indique)

**Pull request merged**